raft/raft.go

```diff
@@ -726,12 +727,19 @@ func (r *raft) Step(m pb.Message) error {
 			// but it will not receive MsgApp or MsgHeartbeat, so it will not create
 			// disruptive term increases
 			r.send(pb.Message{To: m.From, Type: pb.MsgAppResp})
-		} else {
+			return nil
+		case m.Type == pb.MsgPreVote || m.Type == pb.MsgPreVoteResp:
+			// If we receive a MsgPreVote from a node with a lower term number,
+			// we reject it. For a MsgPreVoteResp we simply pass it to our
+			// stepFunc. For a stale MsgPreVoteResp (think of late responses,
+			// we've already transitioned to a leader or follower state) this
```

Comment:
I see. Short of augmenting `pb.Message` I see no solution to this (unless I'm missing something). Say I have a three node cluster (A, B & C) where there is an asymmetric partition between A & C in that it C can _receive_ incoming messages but cannot deliver _outgoing_ ones to A. Consider the following:

- A campaigns, send out a MsgPreVote to C
- C responds with MsgPreVoteResp, doesn't deliver (yet)
- A returns back to follower state (B wins election)
   ...
- Timeout, A campaigns again (later term), sends out MsgPreVote to C
- C responds with _another_ MsgPreVoteResp, doesn't deliver (yet)
[partition heals]
- A receives two MsgPreVoteResp messagess, cannot distinguish which one is in response to which election. (This is because C doesn't change terms when it receives MsgPreVote messages, pb.Message.Term in each MsgPreVoteResp message is simply C's term when it received said message.)

Ignored is the fact that in this specific example with the three node cluster necessarily C's term has increased between the two messages. Point stands where we still have no good way to determine what election cycle each MsgPreVoteResp message is from.